[RCORE-426] Refactor util::Scheduler so multiple implementations can be included

simonask opened the pull request:
## What, How & Why?

The implementation of `util::Scheduler` was hardcoded at compile-time, which meant that multiple implementations could not be compiled into the binary. This prevented reuse of binaries targeting the same platform but different runtimes (example: JS runtimes running on Apple hardware), and it prevented SDKs from implementing a custom scheduler outside of Object Store code.

This PR includes:

* `Scheduler::set_default_factory()` now works on all platforms.
* `Scheduler::get_frozen()` is deprecated in favor of `Scheduler::make_frozen()` (naming).
* The _default_ scheduler implementation is still selected at compile-time, sensitive to compile flags and availability of libraries (`REALM_HAVE_UV`/`REALM_USE_UV`).

https://jira.mongodb.org/browse/RCORE-426

## ☑️ ToDos
* [ ] 📝 Changelog update
* ~~[ ] 🚦 Tests (or not relevant)~~

fealebenpae commented on src/realm/object-store/util/android/scheduler.hpp:
```diff
@@ -34,9 +34,9 @@
 
-namespace {
-using namespace realm;
+namespace realm::util {
```
C++17 FTW!

tgoyne reviewed (commented):
I'm not a fan of the addition of preprocessor soup rather than leaving the functions in the platform-specific files. This also is making a bunch of previously private symbols exported when they don't particularly need to be.

jedelbo reviewed (commented):
I have no position on this change. I will let the consumers of the functionality decide.

simonask commented:
> I'm not a fan of the addition of preprocessor soup rather than leaving the functions in the platform-specific files.

I would be fine with moving the definition of platform-specific `make` functions into separate cpp files. Still need a bit of 
preprocessor soup specifically for Cocoa, though.

> This also is making a bunch of previously private symbols exported when they don't particularly need to be.

Good point.

simonask commented:
I toyed around with addressing your concerns @tgoyne, but I honestly don't think that there is much to be gained from moving things around to minimize preprocessor statements.

I'm also not too worried about exporting more symbols. Is there a reason to be concerned about that?
